src/Common/serde.h

```diff
@@ -0,0 +1,63 @@
+#pragma once
+
+#include <IO/VarInt.h>
+#include <base/types.h>
+#include <Common/VersionRevision.h>
+
+namespace DB
+{
+/// REQUIRES: The object must support versioned serialization/deserialization
+template <typename S, typename WB>
+concept VersionedSerializable = requires(const S & s, WB && wb, VersionType version) { s.serialize(wb, version); };
+
+template <typename S, typename RB>
+concept VersionedDeserializable = requires(S && s, RB && rb, VersionType version) { s.deserialize(rb, version); };
+
+template <typename T>
+concept HasGetVersionFromRevision = requires(T && a) {
+    {
+        a.getVersionFromRevision(ProtonRevision::getVersionRevision())
+    } -> std::convertible_to<VersionType>;
+};
+
+template <typename WB, VersionedSerializable<WB> S>
+void ALWAYS_INLINE serialize(const S & s, WB && wb, std::optional<VersionType> version = std::nullopt)
```

Comment:
This appears too complex and has too many branches. I like to recommend
1) make code simple to understand 
2) make code branch less as much as possible 

Simple code is usually faster to execute